File: dotnet/src/Cryptunnel.Core/Crypto/AesGcmCipher.cs
using System;
using System.Security.Cryptography;

namespace Cryptunnel.Core.Crypto;

/// <summary>
/// AES-256-GCM 隧道加密算法，与 Java 侧 <c>AesGcmCipher</c> 字节级对应。
///
/// GCM 本身是 AEAD，密文尾部自带 16 字节认证标签，故<b>不再外挂 HMAC</b>。
///
/// 载荷格式（比默认算法短 24 字节：省掉 32 字节 HMAC，改用 16 字节标签，nonce 也从 16 减到 12）：
/// <code>
///   Base64( nonce[12] || 密文 || tag[16] )
///   密钥派生：SHA256("AES:" + rawKey) —— 与默认算法一致，同一把 rawKey 可直接复用
/// </code>
///
/// <b>为什么 nonce 是 12 字节</b>：GCM 标准 nonce 长度为 96 bit，
/// 非 12 字节需要额外 GHASH 派生且各实现支持度不一，统一用 12 字节保证三端互通。
///
/// <b>与 Java 侧的一处结构差异需注意</b>：.NET 的 <see cref="AesGcm"/> 把 tag 与密文分开传参，
/// 而 Java 的 <c>Cipher</c> 把 tag 追加在密文尾部。本实现按 Java 的布局拼装，
/// 即 tag 位于载荷末尾，两端才能互解。
/// </summary>
public sealed class AesGcmCipher : ITunnelCipher
{
    /// <summary>算法标识，与 Java 侧 <c>AesGcmCipher.ID</c> 完全一致。</summary>
    public const string CipherId = "aes-256-gcm";

    private const int NonceLength = 12;
    private const int TagLength = 16;

    public string Id => CipherId;

    public string Seal(ReadOnlySpan<byte> plaintext, string rawKey)
    {
        var nonce = RandomNumberGenerator.GetBytes(NonceLength);
        var key = AesCrypto.DeriveAesKey(rawKey);

        // 布局：nonce || 密文 || tag。tag 追加在尾部以匹配 Java 的 Cipher 输出。
        var result = new byte[NonceLength + plaintext.Length + TagLength];
        nonce.CopyTo(result.AsSpan(0, NonceLength));

        using var gcm = new AesGcm(key, TagLength);
        gcm.Encrypt(
            nonce,
            plaintext,
            result.AsSpan(NonceLength, plaintext.Length),
            result.AsSpan(NonceLength + plaintext.Length, TagLength));

        return Convert.ToBase64String(result);
    }

    public byte[] Open(string payload, string rawKey)
    {
        var raw = Convert.FromBase64String(payload);
        if (raw.Length < NonceLength + TagLength)
            throw new CryptographicException(
                $"aes-256-gcm 载荷长度不足：{raw.Length} 字节，至少需要 {NonceLength + TagLength} 字节（nonce+tag）");

        var cipherLength = raw.Length - NonceLength - TagLength;
        var nonce = raw.AsSpan(0, NonceLength);
        var ciphertext = raw.AsSpan(NonceLength, cipherLength);
        var tag = raw.AsSpan(NonceLength + cipherLength, TagLength);

        var plaintext = new byte[cipherLength];
        using var gcm = new AesGcm(AesCrypto.DeriveAesKey(rawKey), TagLength);
        // 密钥错误与数据篡改都表现为 AuthenticationTagMismatchException（CryptographicException 子类），
        // 刻意不区分二者——区分会泄露信息。
        gcm.Decrypt(nonce, ciphertext, tag, plaintext);
        return plaintext;
    }
}

File: java/cryptunnel-core/src/main/java/io/github/objectyan/cryptunnel/core/crypto/AesGcmCipher.java
package io.github.objectyan.cryptunnel.core.crypto;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.SecureRandom;
import java.util.Base64;

/**
 * AES-256-GCM 隧道加密算法（JDK 8 原生支持，无需额外依赖）。
 *
 * <p>GCM 本身是 AEAD（带认证的加密），密文尾部自带 16 字节认证标签，
 * 因此<b>不再需要外挂 HMAC</b>——再加一层 HMAC 只会增加长度和复杂度而不提升安全性。</p>
 *
 * <p>载荷格式（比默认算法更短：省掉 32 字节 HMAC，改用 16 字节 GCM 标签）：</p>
 * <pre>
 *   Base64( IV[12] || AES/GCM/NoPadding 密文（末 16 字节为 GCM 标签） )
 *   标签长度：128 bit
 *   密钥派生：SHA-256("AES:" + rawKey) —— 与默认算法一致，同一把 rawKey 可直接复用
 * </pre>
 *
 * <p><b>为什么 IV 是 12 字节</b>：GCM 的标准 nonce 长度为 96 bit。12 字节以外的长度
 * 需要额外的 GHASH 派生，且各实现支持度不一（.NET 的 {@code AesGcm} 直到 .NET 8
 * 才放开非 12 字节 nonce），统一用 12 字节可保证三端互通。</p>
 *
 * <p><b>注意</b>：AES-256 在 JDK 8u161 之前需要安装 JCE 无限制强度策略文件。
 * 该要求与默认的 {@link AesCbcHmacSha256Cipher} 相同，不是新增约束。</p>
 */
public final class AesGcmCipher implements TunnelCipher {

    /** 算法标识，用于配置匹配；不进报文。 */
    public static final String ID = "aes-256-gcm";

    private static final String TRANSFORMATION = "AES/GCM/NoPadding";
    private static final int IV_LENGTH = 12;
    private static final int TAG_LENGTH_BITS = 128;

    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    @Override
    public String id() {
        return ID;
    }

    @Override
    public String seal(byte[] plaintext, String rawKey) {
        try {
            byte[] iv = new byte[IV_LENGTH];
            SECURE_RANDOM.nextBytes(iv);

            Cipher cipher = Cipher.getInstance(TRANSFORMATION);
            cipher.init(Cipher.ENCRYPT_MODE, deriveKey(rawKey), new GCMParameterSpec(TAG_LENGTH_BITS, iv));
            byte[] ciphertext = cipher.doFinal(plaintext);

            return Base64.getEncoder().encodeToString(AesUtil.concat(iv, ciphertext));
        } catch (Exception e) {
            throw new TunnelCryptoException("aes-256-gcm seal failed: " + e.getMessage(), e);
        }
    }

    @Override
    public byte[] open(String payload, String rawKey) {
        try {
            byte[] raw = Base64.getDecoder().decode(payload);
            if (raw.length < IV_LENGTH + TAG_LENGTH_BITS / 8) {
                throw new TunnelCryptoException(
                        "aes-256-gcm payload too short: " + raw.length + " bytes");
            }

            byte[] iv = AesUtil.subarray(raw, 0, IV_LENGTH);
            byte[] ciphertext = AesUtil.subarray(raw, IV_LENGTH, raw.length);

            Cipher cipher = Cipher.getInstance(TRANSFORMATION);
            cipher.init(Cipher.DECRYPT_MODE, deriveKey(rawKey), new GCMParameterSpec(TAG_LENGTH_BITS, iv));
            return cipher.doFinal(ciphertext);
        } catch (TunnelCryptoException e) {
            throw e;
        } catch (Exception e) {
            // 密钥错误与数据篡改在 GCM 下都表现为 AEADBadTagException，
            // 统一包装，不区分二者（区分会泄露信息）。
            throw new TunnelCryptoException("aes-256-gcm open failed: " + e.getMessage(), e);
        }
    }

    /**
     * 与默认算法共用派生前缀，使同一把 rawKey 在切换算法时无需更换。
     */
    private static SecretKeySpec deriveKey(String rawKey) {
        return AesUtil.deriveAesKey(rawKey);
    }
}
